File: client/src/modules/nexotype/schemas/clinical/phenotype.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const PhenotypeSchema = z.object({
  id: z.number(),
  name: z.string().min(1).max(255),
  hpo_id: z.string().max(20).nullable().optional(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreatePhenotypeSchema = z.object({
  name: z.string().min(1).max(255),
  hpo_id: z.string().max(20).nullable().optional(),
});

export const UpdatePhenotypeSchema = z.object({
  name: z.string().min(1).max(255).optional(),
  hpo_id: z.string().max(20).nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type Phenotype = z.infer<typeof PhenotypeSchema>;
export type CreatePhenotype = z.infer<typeof CreatePhenotypeSchema>;
export type UpdatePhenotype = z.infer<typeof UpdatePhenotypeSchema>;

// ==========================================
// Response Types
// ==========================================

export type PhenotypeResponse = {
  success: boolean;
  data?: Phenotype;
  error?: string;
};

export type PhenotypesResponse = {
  success: boolean;
  data?: Phenotype[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/clinical/phenotype_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class PhenotypeCreate(BaseModel):
    """Schema for creating a new phenotype"""
    name: str
    hpo_id: str | None = None


class PhenotypeUpdate(BaseModel):
    """Schema for updating a phenotype"""
    name: str = None
    hpo_id: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class PhenotypeDetail(BaseModel):
    """Schema for phenotype details"""
    id: int
    name: str
    hpo_id: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class PhenotypeListResponse(BaseModel):
    """Response schema for listing phenotypes"""
    success: bool
    data: list[PhenotypeDetail] | None = None
    count: int = 0
    error: str | None = None


class PhenotypeResponse(BaseModel):
    """Response schema for single phenotype operations"""
    success: bool
    data: PhenotypeDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
